src/webgpu/api/operation/vertex_state/correctness.spec.ts

```diff
@@ -789,3 +747,333 @@ g.test('buffersWithVaryingStepMode')
+    const expectedDataBG = t.createExpectedBG(testData, pipeline);
+    t.submitRenderPass(pipeline, vertexBuffers, expectedDataBG, kVertexCount, kInstanceCount);
+  });
+
+g.test('vertexBufferUsedMultipleTime_interleaved')
+  .desc(
+    `Test using the same vertex buffer in for multiple "vertex buffers", with data from each buffer interleaved.
+  - For each vertex format.
+  - For various numbers of vertex buffers [2, 3, max]`
+  )
+  .params(u =>
+    u //
+      .combine('format', kVertexFormats)
+      .beginSubcases()
+      .combine('vbCount', [2, 3, kMaxVertexBuffers])
+      .combine('additionalVBOffset', [0, 4, 120])
+  )
+  .fn(t => {
+    const { format, vbCount, additionalVBOffset } = t.params;
+    const kVertexCount = 20;
+    const kInstanceCount = 1;
+    const formatInfo = kVertexFormatInfo[format];
+    const formatByteSize = formatInfo.bytesPerComponent * formatInfo.componentCount;
+    // We need to align so the offset for non-0 setVertexBuffer don't fail validation.
+    const alignedFormatByteSize = align(formatByteSize, 4);
+
+    // Create data for a single vertex buffer with many attributes, that will be split between
+    // many vertexbuffers set at different offsets.
+    const attribs: GPUVertexAttribute[] = [];
+    for (let i = 0; i < vbCount; i++) {
+      attribs.push({ format, offset: i * alignedFormatByteSize, shaderLocation: i });
+    }
+    const baseData = t.createTestAndPipelineData(
+      [
+        {
+          slot: 0,
+          arrayStride: alignedFormatByteSize * vbCount,
+          stepMode: 'vertex',
+          vbOffset: additionalVBOffset,
+          attributes: attribs,
+        },
+      ],
+      kVertexCount,
+      kInstanceCount
+    );
+    const vertexBuffer = t.createVertexBuffers(baseData, kVertexCount, kInstanceCount)[0].buffer;
+
+    // Split the 1 vertex buffer with N attributes into N vertex buffers with one attribute.
+    const testData: VertexLayoutState<{}, TestData> = [];
+    const vertexBuffers: VertexState<{ buffer: GPUBuffer; vbOffset: number }, {}> = [];
+    for (let i = 0; i < vbCount; i++) {
+      vertexBuffers.push({
+        slot: i,
+        buffer: vertexBuffer,
+        vbOffset: additionalVBOffset + i * alignedFormatByteSize,
+        attributes: [],
+      });
+      testData.push({
+        ...baseData[0],
+        slot: i,
+        attributes: [{ ...baseData[0].attributes[i], offset: 0 }],
+      });
+    }
+
+    // Run the test with the modified test data.
+    const pipeline = t.makeTestPipeline(testData, kVertexCount, kInstanceCount);
+    const expectedDataBG = t.createExpectedBG(testData, pipeline);
+    t.submitRenderPass(pipeline, vertexBuffers, expectedDataBG, kVertexCount, kInstanceCount);
+  });
+
+g.test('maxBuffersAndAttribs')
+  .desc(
+    `Test a vertex state that loads as many attributes and buffers as possible.
+  - For each format.
+  TODO find a way to test maxAttribs. Right now this test is gated on kMaxUniformBuffersPerStage
```

Comment:
Can we modify the generated shader to read from both uniform buffers and storage buffers?

Reply:
We can. I'd like to do that in a follow-up PR if you don't mind.

Reply:
Sounds fine